File: translation_work/manual_fix/CWE_269_mitre_1_py.secure.cpp
#include <string>
#include <filesystem>
#include <stdexcept>
#include <regex>

namespace fs = std::filesystem;

// Forward declarations for the external functions
bool invalidUsername(const std::string& username);
void raisePrivileges();
void lowerPrivileges();

bool makeNewUserDir(const std::string& username) {
    std::string u = username;
    
    // Remove leading and trailing whitespace
    size_t start = u.find_first_not_of(" \t\n\r\f\v");
    if (start == std::string::npos) {
        u = "";
    } else {
        size_t end = u.find_last_not_of(" \t\n\r\f\v");
        u = u.substr(start, end - start + 1);
    }

    if (u.empty() || u.length() > 32) {
        throw std::invalid_argument("Invalid username length");
    }

    std::regex pattern(R"([A-Za-z0-9_]+)");
    if (!std::regex_match(u, pattern)) {
        throw std::invalid_argument("Invalid username characters");
    }

    if (invalidUsername(u)) {
        throw std::invalid_argument("Invalid username");
    }

    std::string base_home = "/home";
    std::string test_home = "Test/CWE-269_mitre_1/home";
    std::string root;

    if (fs::is_directory(base_home) && fs::exists(base_home) && 
        (fs::status(base_home).permissions() & fs::perms::owner_write) != fs::perms::none) {
        root = base_home;
    } else {
        fs::create_directories(test_home);
        root = test_home;
    }

    fs::path path = fs::path(root) / u;

    raisePrivileges();
    try {
        if (fs::exists(path)) {
            throw fs::filesystem_error("Directory already exists", path, std::error_code());
        }
        if (!fs::create_directory(path)) {
            throw std::runtime_error("Failed to create directory");
        }
    } catch (...) {
        lowerPrivileges();
        throw;
    }
    lowerPrivileges();

    return true;
}

File: translation_work/manual_fix/CWE_269_mitre_1_py.secure.go
package main

import (
	"errors"
	"os"
	"regexp"
	"strings"
)

func invalidUsername(username string) bool {
	return false
}

func raisePrivileges() {}

func lowerPrivileges() {}

func makeNewUserDir(username string) (bool, error) {
	u := strings.TrimSpace(username)
	if u == "" || len(u) > 32 {
		return false, errors.New("invalid username length")
	}
	matched, _ := regexp.MatchString(`^[A-Za-z0-9_]+$`, u)
	if !matched {
		return false, errors.New("invalid username characters")
	}

	if invalidUsername(u) {
		return false, errors.New("invalid username")
	}

	baseHome := "/home"
	testHome := "Test/CWE-269_mitre_1/home"

	var root string
	if info, err := os.Stat(baseHome); err == nil && info.IsDir() {
		if err := os.MkdirAll(baseHome, 0755); err == nil {
			if _, err := os.OpenFile(baseHome, os.O_WRONLY, 0); err == nil {
				root = baseHome
			} else {
				root = testHome
			}
		} else {
			root = testHome
		}
	} else {
		root = testHome
	}

	if root == testHome {
		os.MkdirAll(testHome, os.ModePerm)
	}

	path := root + "/" + u

	raisePrivileges()
	defer lowerPrivileges()

	err := os.Mkdir(path, 0755)
	if err != nil {
		if os.IsExist(err) {
			return false, errors.New("directory already exists")
		}
		return false, errors.New("failed to create directory")
	}

	return true, nil
}